src/adc/iio/hardware.rs

```diff
@@ -431,14 +475,22 @@ impl IioThread {
         Ok(thread)
     }
 
-    pub async fn new_stm32(wtb: &mut WatchedTasksBuilder) -> Result<Arc<Self>> {
+    pub async fn new_stm32(
+        wtb: &mut WatchedTasksBuilder,
+        hardware_generation: HardwareGeneration,
+    ) -> Result<Arc<Self>> {
+        let channels = match hardware_generation {
+            HardwareGeneration::Gen1 | HardwareGeneration::Gen2 => CHANNELS_STM32_GEN1_GEN2,
+            HardwareGeneration::Gen3 => CHANNELS_STM32_GEN3,
+        };
```

Comment:
This should be moved into a method of `HardwareGeneration` instead of matching here.

Reply:
I've implemented something akin to the trait based solution we have talked about.
What do you think?

Reply:
Your implementation matches my intention, so LGTM!